---
title: CA0054
ms.date: 10/19/2016
ms.topic: reference
f1_keywords:
  - "CA0054"
helpviewer_keywords:
  - "CA0054"
ms.assetid: ebc7feed-8c4d-4302-8df1-85450b432cff
author: mikejo5000
ms.author: mikejo
manager: jillfra
ms.workload:
  - "multiple"
---
# CA0054
Custom rule assembly has invalid XML resources

This occurs if a custom rule assembly has invalid XML resources describing its rules. The correct format is as follows:

```xml
 <Rules FriendlyName="Customer Rules">
   <Rule TypeName="[The unqualified type name of the rule]" Category="[A category name such as Customer.Usage]" CheckId="[An identifier for the rule that is at least unique within the same category]">
     <Name>[Localized version of the type name, this is the rule name that appears in the UI]</Name>
     <Description>[A sentence describing the rule in more detail than the name].</Description>
     <Url>[A URL pointing to a documentation or info about the rule]</Url>
     <Resolution>[Format string for rule messages].</Resolution>
     <Email>[Email Address]</Email>
     <MessageLevel Certainty="[A number from 0 to 100]">[Error or Warning]</MessageLevel>
     <FixCategories>[NonBreaking or Breaking]</FixCategories>
     <Owner>[Rule owner name or group]</Owner>
   </Rule>
 </Rules>
```

## See also
[Code Analysis Application Errors](../code-quality/code-analysis-application-errors.md)
