---
title: "CA1304: Specify CultureInfo (code analysis)"
description: "Learn about code analysis rule CA1304: Specify CultureInfo"
ms.date: 06/30/2018
ms.topic: reference
f1_keywords:
- SpecifyCultureInfo
- CA1304
helpviewer_keywords:
- SpecifyCultureInfo
- CA1304
author: gewarren
ms.author: gewarren
---
# CA1304: Specify CultureInfo

|Item|Value|
|-|-|
|RuleId|CA1304|
|Category|Microsoft.Globalization|
|Fix is a breaking or non-breaking change|Non-breaking|

## Cause

A method or constructor calls a member that has an overload that accepts a <xref:System.Globalization.CultureInfo?displayProperty=nameWithType> parameter, and the method or constructor does not call the overload that takes the <xref:System.Globalization.CultureInfo> parameter. This rule ignores calls to the following methods:

- <xref:System.Activator.CreateInstance%2A?displayProperty=nameWithType>
- <xref:System.Resources.ResourceManager.GetObject%2A?displayProperty=nameWithType>
- <xref:System.Resources.ResourceManager.GetString%2A?displayProperty=nameWithType>

You can also [configure](#configurability) more symbols to be excluded by this rule.

## Rule description

When a <xref:System.Globalization.CultureInfo> or <xref:System.IFormatProvider?displayProperty=nameWithType> object is not supplied, the default value that is supplied by the overloaded member might not have the effect that you want in all locales. Also, .NET members choose default culture and formatting based on assumptions that might not be correct for your code. To ensure the code works as expected for your scenarios, you should supply culture-specific information according to the following guidelines:

- If the value will be displayed to the user, use the current culture. See <xref:System.Globalization.CultureInfo.CurrentCulture%2A?displayProperty=nameWithType>.

- If the value will be stored and accessed by software, that is, persisted to a file or database, use the invariant culture. See <xref:System.Globalization.CultureInfo.InvariantCulture%2A?displayProperty=nameWithType>.

- If you do not know the destination of the value, have the data consumer or provider specify the culture.

Even if the default behavior of the overloaded member is appropriate for your needs, it is better to explicitly call the culture-specific overload so that your code is self-documenting and more easily maintained.

> [!NOTE]
> <xref:System.Globalization.CultureInfo.CurrentUICulture%2A?displayProperty=nameWithType> is used only to retrieve localized resources by using an instance of the <xref:System.Resources.ResourceManager?displayProperty=nameWithType> class.

## How to fix violations

To fix a violation of this rule, use the overload that takes a <xref:System.Globalization.CultureInfo> argument.

## When to suppress warnings

It is safe to suppress a warning from this rule when it is certain that the default culture is the correct choice, and where code maintainability is not an important development priority.

## Configurability

This rule has the following configurable options.

### Excluded symbol names

You can configure which parts of your codebase to exclude from analysis. For example, to specify that the rule should not run on any code within types named `MyType`, add the following key-value pair to an .editorconfig file in your project:

```ini
dotnet_code_quality.CA1304.excluded_symbol_names = MyType
```

Allowed symbol name formats in the option value (separated by '|'):

- Symbol name only (includes all symbols with the name, regardless of the containing type or namespace)
- Fully qualified names in the symbol's [documentation ID format](https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md#id-string-format). Each symbol name requires a symbol kind prefix, such as "M:" prefix for methods, "T:" prefix for types, "N:" prefix for namespaces, etc.
- `.ctor` for constructors and `.cctor` for static constructors

Examples:

| Option Value | Summary |
| --- | --- |
|`dotnet_code_quality.CA1304.excluded_symbol_names = MyType` | Matches all symbols named 'MyType' in the compilation
|`dotnet_code_quality.CA1304.excluded_symbol_names = MyType1|MyType2` | Matches all symbols named either 'MyType1' or 'MyType2' in the compilation
|`dotnet_code_quality.CA1304.excluded_symbol_names = M:NS.MyType.MyMethod(ParamType)` | Matches specific method 'MyMethod' with given fully qualified signature
|`dotnet_code_quality.CA1304.excluded_symbol_names = M:NS1.MyType1.MyMethod1(ParamType)|M:NS2.MyType2.MyMethod2(ParamType)` | Matches specific methods 'MyMethod1' and 'MyMethod2' with respective fully qualified signature

You can configure all of these options for just this rule, for all rules, or for all rules in this category (Security). For more information, see [Configure code analysis rules](../configure-rules.md).

## Example showing how to fix violations

In the following example, `BadMethod` causes two violations of this rule. `GoodMethod` corrects the first violation by passing the invariant culture to <xref:System.String.Compare%2A?displayProperty=nameWithType>, and corrects the second violation by passing the current culture to <xref:System.String.ToLower%2A?displayProperty=nameWithType> because `string3` is displayed to the user.

[!code-csharp[FxCop.Globalization.CultureInfo#1](../../../../samples/snippets/fundamentals/code-analysis/csharp/ca1304-specify-cultureinfo_1.cs)]

## Example showing formatted output

The following example shows the effect of current culture on the default <xref:System.IFormatProvider> that is selected by the <xref:System.DateTime> type.

[!code-csharp[FxCop.Globalization.IFormatProvider#1](../../../../samples/snippets/fundamentals/code-analysis/csharp/ca1304-specify-cultureinfo_2.cs)]

This example produces the following output:

```txt
6/4/1900 12:15:12 PM
06/04/1900 12:15:12
```

## Related rules

- [CA1305: Specify IFormatProvider](ca1305.md)

## See also

- [Using the CultureInfo Class](../../../standard/globalization-localization/globalization.md#work-with-culture-specific-settings)
