---
title: "CA1711: Identifiers should not have incorrect suffix (code analysis)"
description: "Learn about code analysis rule CA1711: Identifiers should not have incorrect suffix"
ms.date: 07/21/2020
ms.topic: reference
f1_keywords:
- CA1711
- IdentifiersShouldNotHaveIncorrectSuffix
helpviewer_keywords:
- CA1711
- IdentifiersShouldNotHaveIncorrectSuffix
author: gewarren
ms.author: gewarren
---
# CA1711: Identifiers should not have incorrect suffix

|Item|Value|
|-|-|
|RuleId|CA1711|
|Category|Microsoft.Naming|
|Fix is a breaking or non-breaking change|Breaking|

## Cause

An identifier has an incorrect suffix.

By default, this rule only looks at externally visible identifiers, but this is [configurable](#configurability).

## Rule description

By convention, only the names of types that extend certain base types or that implement certain interfaces, or types derived from these types, should end with specific reserved suffixes. Other type names should not use these reserved suffixes.

The following table lists the reserved suffixes and the base types and interfaces with which they are associated.

|Suffix|Base type/Interface|
|------------|--------------------------|
|Attribute|<xref:System.Attribute?displayProperty=fullName>|
|Collection|<xref:System.Collections.ICollection?displayProperty=fullName><br/></br/><xref:System.Collections.IEnumerable?displayProperty=fullName><br/></br/><xref:System.Collections.Queue?displayProperty=fullName><br/></br/><xref:System.Collections.Stack?displayProperty=fullName><br/></br/><xref:System.Collections.Generic.ICollection%601?displayProperty=fullName><br/></br/><xref:System.Data.DataSet?displayProperty=fullName><br/></br/><xref:System.Data.DataTable?displayProperty=fullName>|
|Dictionary|<xref:System.Collections.IDictionary?displayProperty=fullName><br/></br/><xref:System.Collections.Generic.IDictionary%602?displayProperty=fullName>|
|EventArgs|<xref:System.EventArgs?displayProperty=fullName>|
|EventHandler|An event-handler delegate|
|Exception|<xref:System.Exception?displayProperty=fullName>|
|Flag|<xref:System.Enum?displayProperty=fullName>|
|Flags|<xref:System.Enum?displayProperty=fullName>|
|Permission|<xref:System.Security.IPermission?displayProperty=fullName>|
|Queue|<xref:System.Collections.Queue?displayProperty=fullName>|
|Stack|<xref:System.Collections.Stack?displayProperty=fullName>|
|Stream|<xref:System.IO.Stream?displayProperty=fullName>|

In addition, the following suffixes should **not** be used:

- `Delegate`

- `Enum`

- `Impl` (use `Core` instead)

- `Ex` or similar suffix to distinguish it from an earlier version of the same type

Naming conventions provide a common look for libraries that target the common language runtime. This reduces the learning curve that is required for new software libraries, and increases customer confidence that the library was developed by someone who has expertise in developing managed code.

## How to fix violations

Remove the suffix from the type name.

## When to suppress warnings

Do not suppress a warning from this rule unless the suffix has an unambiguous meaning in the application domain.

## Configurability

This rule has the following configurable options.

### API surface

You can configure which parts of your codebase to run this rule on, based on their accessibility. For example, to specify that the rule should run only against the non-public API surface, add the following key-value pair to an *.editorconfig* file in your project:

```ini
dotnet_code_quality.ca1711.api_surface = private, internal
```

You can configure this option for just this rule, for all rules, or for all rules in this category (Naming). For more information, see [Configure code analysis rules](../configure-code-quality-rules.md).

### Allow suffixes

You can configure a list of allowed suffixes, with each suffix separated by the pipe character ("|"). For example, to specify that the rule should not run against Flag and Flags suffixes, add the following key-value pair to an *.editorconfig* file in your project:

```ini
dotnet_code_quality.ca1711.allowed_suffixes = Flag|Flags
```

## Related rules

- [CA1710: Identifiers should have correct suffix](ca1710.md)

## See also

- [Attributes](../../../standard/design-guidelines/attributes.md)
- [Handling and raising events](../../../standard/events/index.md)
